pkg/protocols/grpcprot/request.go

```diff
@@ -0,0 +1,298 @@
+// modify its content. The function panic if the payload is a stream.
+func (r *Request) RawPayload() []byte {
+	panic("the payload is a large one")
+}
+
+// PayloadSize returns the size of the payload. If the payload is a
+// stream, it returns the bytes count that have been currently read
+// out.
+func (r *Request) PayloadSize() int64 {
+	panic("implement me")
+}
+
+// ToBuilderRequest wraps the request and returns the wrapper, the
+// return value can be used in the template of the Builder filters.
+func (r *Request) ToBuilderRequest(name string) interface{} {
+	panic("implement me")
+}
+
+// Close returns the request
+func (r *Request) Close() {
+	r.cancel()
+}
+
+// Context returns the context of the request
+func (r *Request) Context() context.Context {
+	return r.ctx
+}
+
+// FullMethod returns full method name of the grpc request.
+// The returned string is in the format of "/service/method"
+func (r *Request) FullMethod() string {
+	return r.sts.path
+}
+
+// Method returns method name of the grpc request.
+func (r *Request) Method() string {
+	if index := strings.LastIndex(r.sts.path, "/"); index != -1 {
+		return r.sts.path[index+1:]
+	}
+	return r.sts.path
+}
+
+// Service returns service name of the grpc request.
+func (r *Request) Service() string {
+	if index := strings.LastIndex(r.sts.path, "/"); index != -1 {
+		return r.sts.path[0:index]
+	}
+	return r.sts.path
+}
+
+// SetFullMethod sets full method name o
+//f the grpc request.
+func (r *Request) SetFullMethod(path string) {
+	r.sts.setMethod(path)
+}
+
+// RealIP return the client ip from peer.FromContext()
+func (r *Request) RealIP() string {
+	ip := r.realIP
+	if ip != "" {
+		return ip
+	}
+	if strings.ContainsRune(r.peer.Addr.(*Addr).String(), ':') {
+		ip, _, _ = net.SplitHostPort(r.peer.Addr.(*Addr).String())
+	} else {
+		ip = r.peer.Addr.(*Addr).String()
+	}
+	r.realIP = ip
+	return ip
+}
+
+func (r *Request) SetRealIP(ip string) {
+	r.realIP = ""
+	if strings.ContainsRune(r.peer.Addr.String(), ':') {
+		_, port, _ := net.SplitHostPort(r.peer.Addr.String())
+		r.peer.Addr.(*Addr).setAddr(fmt.Sprintf("%s:%s", ip, port))
+	} else {
+		r.peer.Addr.(*Addr).setAddr(ip)
+	}
+}
+
+// SourceHost returns peer host from grpc context
+func (r *Request) SourceHost() string {
+	return r.peer.Addr.(*Addr).String()
+}
+
+func (r *Request) SetSourceHost(sourceHost string) {
+	r.peer.Addr.(*Addr).setAddr(sourceHost)
+}
+
+// Host refer to google.golang.org\grpc@v1.46.2\internal\transport\http2_server.go#operateHeaders
+// ensure only one value
+func (r *Request) Host() string {
+	return r.GetFirstInHeader(Authority)
+}
+
+func (r *Request) SetHost(host string) {
+	r.header.Set(Authority, host)
+}
+
+// GetFirstInHeader returns the first value for a given key.
+// k is converted to lowercase before searching in md.
+func (r *Request) GetFirstInHeader(k string) string {
+	v := r.header.md.Get(k)
+	if v == nil {
+		return ""
+	}
+	return v[0]
+}
```

Comment:
the value structure in the grpc header is []string, so sometimes it will be necessary to obtain the first one, for reference 
 [scg](https://github.com/spring-cloud/spring-cloud-gateway/blob/fdd348a1045db182661e90eaf6fd351e5f2b21de/spring-cloud-gateway-server/src/main/java/org/springframework/cloud/gateway/filter/WebsocketRoutingFilter.java#L136) . it will be better to use `r.RawHeader().GetFirst(k)`